stdlib/public/core/DiscontiguousSlice.swift

```diff
@@ -0,0 +1,451 @@
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+/// A collection wrapper that provides access to the elements of a collection,
+/// indexed by a set of indices.
+@available(SwiftStdlib 5.11, *)
+public struct DiscontiguousSlice<Base: Collection> {
+  internal var _base: Base
+  
+  /// The set of subranges that are available through this discontiguous slice.
+  public let subranges: RangeSet<Base.Index>
+
+  /// The collection that the indexed collection wraps.
+  public var base: Base {
+    return _base
+  }
+
+  internal init(_base: Base, subranges: RangeSet<Base.Index>) {
+    self._base = _base
+    self.subranges = subranges
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice: Sendable
+where Base: Sendable, Base.Index: Sendable {}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice: Equatable where Base.Element: Equatable {
+  public static func ==(lhs: Self, rhs: Self) -> Bool {
+    lhs.elementsEqual(rhs)
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice: Hashable where Base.Element: Hashable {
+  public func hash(into hasher: inout Hasher) {
+    var count = 0
+    for element in self {
+      hasher.combine(element)
+      count += 1
+    }
+    hasher.combine(count) // discriminator
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice: CustomStringConvertible {
+  public var description: String {
+    _makeCollectionDescription()
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice {
+  /// A position in a `DiscontiguousSlice`.
+  public struct Index {
+    /// The index of the range that contains `base`.
+    internal let _rangeOffset: Int
+    
+    /// The position of this index in the base collection.
+    public let base: Base.Index
+
+    internal init(_rangeOffset: Int, base: Base.Index) {
+      self._rangeOffset = _rangeOffset
+      self.base = base
+    }
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice.Index: Equatable {
+  public static func == (left: Self, right: Self) -> Bool {
+    left.base == right.base
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice.Index: Hashable where Base.Index: Hashable {
+  public func hash(into hasher: inout Hasher) {
+    hasher.combine(base)
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice.Index: Comparable {
+  public static func < (left: Self, right: Self) -> Bool {
+    left.base < right.base
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice.Index: CustomStringConvertible {
+  public var description: String {
+    var d = ""
+    debugPrint(base, terminator: "", to: &d)
+    d += " (rangeOffset: \(_rangeOffset))"
+    return d
```

Comment:
This will read a bit weird for common `base` indices, such as `Array`'s `Int` indices, or `String.Index` (` (rangeOffset:)

```
Array:
    42 (rangeOffset: 2)
String: (current, potential future)
    Index(_rawBits: 262407) (rangeOffset: 2)
    4[utf8] (rangeOffset: 2)
```

I suggest putting the components in some explicitly delimited list.

```suggestion
    "<base: \(String(reflecting: base), rangeOffset: \(_rangeOffset)>"
```

Proposed result:

```
Array:
    <base: 42, rangeOffset: 2>
String: (current, potential future)
    <base: Index(_rawBits: 262407), rangeOffset: 2>
    <base: 4[utf8], rangeOffset: 2>
```

We do not need to spell out the type name in descriptions (it actually interferes with readability of bulk prints), but the formats we return should generally have some syntactic structure.

The boring alternative would be to print this as `Index(base: 42, rangeOffset: 2)`, similar to how mirror-based descriptions work. I thing shorter descriptions are sweeter, though, and printing `Index` isn't really that useful.

Alternatively, we could also just print `base` as is, omitting rangeOffset from the description altogether (as it is an internal detail that may not be very relevant while debugging.)

Reply:
Sounds good, I like the proposed description you have there so I'll update to use that one